### ISSUE TITLE
multi-interlock config support on single UCP cluster

### ISSUE THREAD

**Issue by nicolaka** [ehazlett/interlock#89]
Allow having multiple/different interlock configs per UCP cluster. Potentially change key from `interlock/v1/config` to something that adds an interlock identifier ?

**Comment by everett-toews** [comment]
I have this requirement too. Modern applications tend to be composed of multiple distinct services. Each one of those services needs its own proxy.

Currently it seems like Interlock writes the same config to all proxies of the same type (e.g. nginx). Looking at [the code for identifying proxies](https://github.com/ehazlett/interlock/blob/overlay-support/ext/lb/lb.go#L246-L251), I don't see any other options but maybe I'm missing something.

Either way we need to be able to have multiple Interlock instances each managing a single proxy (i.e. one to one) or have one Interlock instance managing multiple _distinct_ proxies (i.e. one to many).

@ehazlett Thoughts on this feature?

**Comment by everett-toews** [comment]
Proposal for a one to many design...

`config.toml` adds a config option for the service name. e.g.

```
ListenAddr = ":8080"
DockerURL = "tcp://192.168.99.105:3376"
TLSCACert = "/etc/docker/ca.pem"
TLSCert = "/etc/docker/server.pem"
TLSKey = "/etc/docker/server-key.pem"
AllowInsecure = false
EnableMetrics = true

[[extensions]]
name = "nginx"
serviceName="customer"
configPath = "/etc/nginx/nginx.conf"
pidPath = "/tmp/nginx.pid"
templatePath = "/etc/interlock/nginx.conf.template"
maxConn = 1024
port = 80

[[extensions]]
name = "nginx"
serviceName="order"
configPath = "/etc/nginx/nginx.conf"
pidPath = "/tmp/nginx.pid"
templatePath = "/etc/interlock/nginx.conf.template"
maxConn = 1024
port = 80
```

If the `serviceName` is present, Interlock would group proxies by `serviceName`.

Containers that want to participate in a service would identify themselves with the `interlock.ext.service.name` label. e.g.

```
docker run --detach \
  --net mynetwork \
  --expose 8080 \
  --label interlock.network=mynetwork \
  --label interlock.ext.service.name=order \
  acme/order-service:1.9
```

**Comment by ehazlett** [comment]
I understand the request but I'm not sure on the proposal.  This means that each Interlock instance would load up all configuration and be responsible for loading its config.  What happens if no service name is presented?  What happens when the config is updates -- why should all instances need to reload each others configuration?

**Comment by everett-toews** [comment]
I think I understand your concerns. And I think they revolve around the one to many nature of the design. I'll do my best to answer.

> What happens if no service name is presented?

Interlock would continue to act as it does now.

> What happens when the config is updated -- why should all instances need to reload each others configuration?

Let me restate this using a concrete example using the `config.toml` above to see if I understand it. Please correct me if I'm misunderstanding.

If a container with `--label interlock.ext.service.name=order` is created, Interlock would restart all proxies for all services. That's because Interlock does not store any state so it can't decide what proxies should be restarted and what proxies shouldn't be started.

This could potentially be solved by using the `since` parameter in [List Containers](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#list-containers) but that would require storing the id/name of the last container seen. That's doable but let's consider also consider a one to one design.

**Comment by everett-toews** [comment]
### One Interlock. One proxy.

`config.toml` adds a config option for the service name. e.g.

```
ListenAddr = ":8080"
DockerURL = "tcp://192.168.99.105:3376"
TLSCACert = "/etc/docker/ca.pem"
TLSCert = "/etc/docker/server.pem"
TLSKey = "/etc/docker/server-key.pem"
AllowInsecure = false
EnableMetrics = true

[[extensions]]
name = "nginx"
configPath = "/etc/nginx/nginx.conf"
serviceName="customer"
pidPath = "/tmp/nginx.pid"
templatePath = "/etc/interlock/nginx.conf.template"
maxConn = 1024
port = 80
```

If the `serviceName` is present, Interlock would group proxies by `serviceName`.

If the `serviceName` is not present, Interlock would continue to act as it does now.

Containers that want to participate in a service would identify themselves with the `interlock.ext.service.name` label. e.g.

```
docker run --detach \
  --net mynetwork \
  --expose 8080 \
  --label interlock.network=mynetwork \
  --label interlock.ext.service.name=customer \
  acme/customer-service:1.9
```

If a container is started or stopped that belongs to a particular service, only the proxy for that service would be reloaded.

If you want another service, you run another instance of Interlock with a different `serviceName`

**Comment by Beastie71** [comment]
Looks like from the comments this will make it into 1.2 is that correct?

**Comment by ehazlett** [comment]
Yes once it's finished review the PR will be merged and available in 1.2.
